wandb/sdk/service/service_base.py

```diff
@@ -0,0 +1,69 @@
+"""grpc service.
+
+Reliably launch and connect to grpc process.
+"""
+
+from abc import abstractmethod
+import datetime
+import enum
+from typing import Any, Dict
+from typing import TYPE_CHECKING
+
+from wandb.proto import wandb_server_pb2 as spb
+from wandb.sdk.wandb_settings import Settings
+
+if TYPE_CHECKING:
+    from google.protobuf.internal.containers import MessageMap
+
+
+def _pbmap_apply_dict(
+    m: "MessageMap[str, spb.SettingsValue]", d: Dict[str, Any]
+) -> None:
+    for k, v in d.items():
+        if isinstance(v, datetime.datetime):
+            continue
+        if isinstance(v, enum.Enum):
+            continue
+        sv = spb.SettingsValue()
+        if v is None:
+            sv.null_value = True
+        elif isinstance(v, int):
+            sv.int_value = v
+        elif isinstance(v, float):
+            sv.float_value = v
+        elif isinstance(v, str):
+            sv.string_value = v
+        elif isinstance(v, bool):
+            sv.bool_value = v
+        elif isinstance(v, tuple):
+            sv.tuple_value.string_values.extend(v)
+        m[k].CopyFrom(sv)
+
+
+class ServiceInterface:
```

Comment:
both `GRPC` and `Sock` share the same construction of the request message, maybe an idea for refactor is to move the commom logic in to here and just have an abstarctmethod for the "infrom" part?

Reply:
agreed, but it seemed small enough (1-2 lines per 4 methods).  I think this is an area for improvement, but it isnt clear if there are cleaner ways to do this that will be more obvious as we get further down this path.  figured for now it was better to have a little code duplication rather than yet another abstraction :).

Reply:
yeah, i think it's a reasonable tradeoff. Thanks!